builtin/credential/chef-node/path_login.go

```diff
@@ -0,0 +1,435 @@
+	if err != nil {
+		return nil, fmt.Errorf("coulnd't retrieve current policy list")
+	}
+
+	if !policyutil.EquivalentPolicies(policies, req.Auth.Policies) {
+		return nil, fmt.Errorf("policies have changed, not renewing")
+	}
+
+	return framework.LeaseExtend(0, 0, b.System())(req, d)
+}
+
+func constructAuthorization(h http.Header) string {
+	authHeaders := make(map[string]string)
+	var keys []string
+	var ret bytes.Buffer
+
+	for k, v := range h {
+		if strings.HasPrefix(k, "X-Ops-Authorization-") {
+			authHeaders[k] = v[0]
+			keys = append(keys, k)
+		}
+	}
+	sort.Strings(keys)
+	for _, v := range keys {
+		ret.WriteString(authHeaders[v])
+	}
+	return ret.String()
+}
+
+func (b *backend) getNodePolicies(req *logical.Request, node string) ([]string, error) {
+	nodeInfo, err := b.retrieveNodeInfo(req, node)
+	if err != nil {
+		return nil, err
+	}
+
+	var clientPols []string
+	clientEntry, err := b.Client(req.Storage, node)
+	if err != nil {
+		return nil, err
+	}
+	if clientEntry != nil {
+		clientPols = clientEntry.Policies
+	}
+
+	var envPols []string
+	envEntry, err := b.Environment(req.Storage, nodeInfo.ChefEnv)
+	if err != nil {
+		return nil, err
+	}
+	if envEntry != nil {
+		envPols = envEntry.Policies
+	}
+
+	var rolePols []string
+	roles := nodeInfo.AutoAttrs.Roles
+	for _, v := range roles {
+		roleEntry, err := b.Role(req.Storage, v)
+		if err != nil {
+			return nil, err
+		}
+		if roleEntry != nil {
+			rolePols = append(rolePols, roleEntry.Policies...)
+		}
+	}
+
+	var tagPols []string
+	tags := nodeInfo.NormalAttrs.Tags
+	for _, v := range tags {
+		tagEntry, err := b.Tag(req.Storage, v)
+		if err != nil {
+			return nil, err
+		}
+		if tagEntry != nil {
+			tagPols = append(tagPols, tagEntry.Policies...)
+		}
+	}
+
+	var allPol []string
+	allPol = append(allPol, clientPols...)
+	allPol = append(allPol, envPols...)
+	allPol = append(allPol, rolePols...)
+	allPol = append(allPol, tagPols...)
+	allPol = strutil.RemoveDuplicates(allPol)
+
+	return allPol, nil
+}
+
+func (b *backend) retrieveNodeInfo(req *logical.Request, targetName string) (*nodeResponse, error) {
+	config, err := b.Config(req.Storage)
+	if err != nil {
+		return nil, err
+	}
+
+	nodeURL, err := url.Parse(config.BaseURL + "/nodes/" + targetName)
+	if err != nil {
+		return nil, err
+	}
+
+	headers, err := authHeaders(config, nodeURL, "GET", nil, true)
+	if err != nil {
+		return nil, err
+	}
+
+	nodeReq, err := http.NewRequest("GET", nodeURL.String(), nil)
+	if err != nil {
+		return nil, err
+	}
+
+	nodeReq.Header = headers
+	client := &http.Client{}
+	resp, err := client.Do(nodeReq)
+	if err != nil {
+		return nil, err
+	}
+
+	defer resp.Body.Close()
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	var jb nodeResponse
+	if err := json.Unmarshal(body, &jb); err != nil {
+		return nil, err
+	}
+
+	return &jb, nil
+}
+
+func (b *backend) retrievePubKey(req *logical.Request, targetName string) (*rsa.PublicKey, error) {
```

Comment:
This will only retrieve the first key for a client, you need to use the https://docs.chef.io/api_chef_server.html#clients-client-keys API to see other keys.